Extension to `RRTMGP.jl` and implement RadiativeTransferModel (#176)
NumericalEarth/Breeze.jl: docs/src/radiative_transfer.md

                                   thermodynamics = :LiquidIcePotentialTemperature)

# Create gray radiation model
optical_thickness = GrayOpticalThicknessOGorman2008(eltype(grid))
radiation = RadiativeTransferModel(grid, constants, optical_thickness;
                                   surface_temperature,
                                   surface_emissivity = 0.98,
                                   surface_albedo = 0.1,
                                   solar_constant = 1361) # W/m²

# Create atmosphere model with DateTime clock for solar position
clock = Clock(time=DateTime(2024, 9, 27, 16, 0, 0))
model = AtmosphereModel(grid; clock, formulation, radiation)
```

When a `DateTime` clock is used, the solar zenith angle is computed automatically from the time and grid location (longitude and latitude).

### Gray Radiation Model

The [`RadiativeTransferModel`](@ref) model computes:

- **Longwave radiation**: Both upwelling and downwelling thermal radiation using RRTMGP's two-stream solver
- **Shortwave radiation**: Direct beam solar radiation

The gray atmosphere optical thickness follows the parameterization by [OGormanSchneider2008](@citet):

```math

> τ_{lw} = α \frac{Δp}{p} \left[ f_l \frac{p}{p_0} + 4 (1 - f_l) \left(\frac{p}{p_0}\right)^4 \right] \left[ τ_e + (τ_p - τ_e) \sin^2 φ \right]

Author: Ok I see it. It is written confusingly. The whole expression is divided by $p$, so you have to notice that because (again confusingly $\sigma = p / p_0$), we get

$$ \frac{f_l \sigma + 4 (1 - f_l) \sigma^4}{p} = \frac{1}{p_0} \left ( f_l + 4 (1 - f_l) \sigma^3 \right ) $$

If i were to write equation 6, I would write it

$$ \tau = \alpha \frac{\Delta p}{p_0} \left [ f_l + 4 (1 - f_l) \left ( \frac{p}{p_0} \right )^3 \right ] \left ( \cdots \right ) $$

alternatively we can define $\Delta \sigma$ and differentiate wrt to $\sigma$ rather than $p$. One of the other.

                          solar_constant = 1361)        # W/m²

# Create atmosphere model with DateTime clock for solar position
clock = Clock(time=DateTime(2024, 9, 27, 16, 0, 0))
model = AtmosphereModel(grid; clock, formulation, radiation)
```

When a `DateTime` clock is used, the solar zenith angle is computed automatically from the time and grid location (longitude and latitude).

### Gray Radiation Model

The [`GrayRadiation`](@ref) model computes:

- **Longwave radiation**: Both upwelling and downwelling thermal radiation using RRTMGP's two-stream solver
- **Shortwave radiation**: Direct beam solar radiation (no scattering) using the O'Gorman optical thickness

The gray atmosphere optical thickness follows the parameterization in [OGormanSchneider2008](@cite):

```math
τ_{lw} = α \frac{Δp}{p} \left( f_l σ + (1-f_l) 4σ^4 \right) \left( τ_e + (τ_p - τ_e) \sin^2 φ \right)
```

where ``σ = p/p_0`` is the normalized pressure, ``φ`` is latitude, and ``α``, ``f_l``, ``τ_e``, ``τ_p`` are empirical parameters.

For shortwave:
```math
τ_{sw} = 2 τ_0 \frac{p}{p_0} \frac{Δp}{p_0}
```

where ``τ_0 = 0.22`` is the shortwave optical depth parameter.

### Radiative Fluxes

> After running [`set!`](@ref) or [`update_state!`](@ref), the radiative fluxes are available from the radiation model:

Reviewer: This is still there

```julia
# Longwave fluxes (ZFaceFields)
F_lw_up = radiation.upwelling_longwave_flux
F_lw_dn = radiation.downwelling_longwave_flux

# Shortwave flux (direct beam only for non-scattering solver)
F_sw = radiation.downwelling_shortwave_flux
```

!!! note "Shortwave Radiation"
    The gray atmosphere uses a non-scattering shortwave approximation, so only
    the direct beam flux is computed. There is no diffuse shortwave or upwelling
    shortwave in this model.

### Solar Zenith Angle

When using a `DateTime` clock, the solar zenith angle is computed from:
- Grid location (longitude from `x`, latitude from `y` for single-column grids)
- Date and time from `model.clock.time`

The calculation accounts for:
- Day of year (for solar declination)
- Hour angle (based on solar time)
- Latitude (for observer position)

## Surface Properties

> The [`GrayRadiativeTransferModel`](@ref) model requires surface properties:

Author: is that because the docstring is in the extension? trying to figure out how to design this, i think i need to ask the user to provide the optical_thickness which is an RRTMGP object

> The [`GrayRadiativeTransferModel`](@ref) model requires surface properties:

Author: ok ill remove this for now and we can iterate in future PRs

# Radiative Transfer

Breeze.jl integrates with [RRTMGP.jl](https://github.com/NumericalEarth/RRTMGP.jl) to provide radiative transfer capabilities for atmospheric simulations. The radiative transfer model computes longwave and shortwave radiative fluxes and heating rates, which are then incorporated into the moist static energy tendency equation.

## Overview

The radiative transfer implementation in Breeze uses a column-based approach, where each horizontal column of the 3D grid is treated independently. This allows efficient computation of radiative fluxes using RRTMGP's optimized column radiation solver.

## Basic Usage

To use radiative transfer in a Breeze simulation, create a `RadiativeTransferModel` and pass it to the `AtmosphereModel` constructor:

```julia
using Breeze
using Oceananigans

# Create grid

> grid = RectilinearGrid(CPU(), Float64; size=(32, 32, 64), x=(0, 10_000), y=(0, 10_000), z=(0, 20_000))

Reviewer 2: ```suggestion
grid = RectilinearGrid(size=(32, 32, 64), x=(0, 10_000), y=(0, 10_000), z=(0, 20_000))
```

# Create radiative transfer model
rtm = RadiativeTransferModel(
    grid;
    surface_emissivity = 0.98,
    surface_albedo_direct = 0.1,
    surface_albedo_diffuse = 0.1,
    cos_zenith = 0.5,
    toa_solar_flux = 1360.0,
    toa_longwave_flux = 0.0
)

# Create atmosphere model with radiative transfer
model = AtmosphereModel(
    grid;
    radiative_transfer = rtm
)
```

## Column Model Example

Here we demonstrate a simple column model calculation of radiative fluxes. This example sets up a single-column atmosphere and computes radiative heating rates.

```julia
using Breeze
using Oceananigans
using CairoMakie

> # Create a single-column grid (1x1 horizontal, 64 vertical levels)

Reviewer 2: ```suggestion
# Create a single-column grid with 64 vertical levels
```

grid = RectilinearGrid(CPU(), Float64; 
    size=(1, 1, 64), 
    x=(0, 1), 
    y=(0, 1), 
    z=(0, 20_000)

> )

Reviewer 2: ```suggestion
grid = RectilinearGrid(; size=64, z=(0, 20_000), topology = (Flat, Flat, Bounded))
```
